### PR TITLE
Chapter1/Homework5

### PR DESCRIPTION
**Pull request by ygnacio**
Implemented array operations: 
1) Sort array in ascending order
2) Inverse array
3) Remove first element from array
4) Remove last element from array
5) Remove element at a given index from array
6) Insert element at the start of array
7) Insert element at the end of array
8) Insert element at specified index of array

### CODE REVIEW

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 5
           

Complexity increasing per file
==============================
- Src/BootCamp.Chapter1/ArrayOperations.cs  8
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4930349&bid=16308082)

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 1]
```diff
@@ -51,8 +151,30 @@ public static int[] RemoveFirst(int[] array)
...
+                return array;
+            }
+
+            if (ArrayIsNullOrEmpty(array))
+            {
+                return array;
+            }
+
+            int j = 0;
+            int[] auxArray = new int[array.Length - 1];
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Codacy Issue found: ['array' is null on at least one execution path.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4930349)

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 2]
```diff
@@ -88,8 +229,41 @@ public static int[] InsertLast(int[] array, int number)
...
+            {
+                if (j == index)
+                {
+                    auxArray[j] = number;
+                    j++;
+                    continue;
+                }
+                auxArray[j] = array[k];
+                j++;
+                k++;
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Codacy Issue found: [Do not update the loop counter 'k' within the loop body.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4930349)

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 3]
```diff
@@ -88,8 +229,41 @@ public static int[] InsertLast(int[] array, int number)
...
+            for (int i = 0, j = 0, k = 0; i < auxArray.Length; i++)
+            {
+                if (j == index)
+                {
+                    auxArray[j] = number;
+                    j++;
+                    continue;
+                }
+                auxArray[j] = array[k];
+                j++;
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Codacy Issue found: [Do not update the loop counter 'j' within the loop body.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4930349)

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 4]
```diff
@@ -88,8 +229,41 @@ public static int[] InsertLast(int[] array, int number)
...
+            if (ArrayIsNullOrEmpty(array))
+            {
+                return new int[] { number };
+            }
+            
+            if (index == 0)
+            {
+                return InsertFirst(array, number);
+            }
+            if (index == array.Length)
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Codacy Issue found: ['array' is null on at least one execution path.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4930349)

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 5]
```diff
@@ -88,8 +229,41 @@ public static int[] InsertLast(int[] array, int number)
...
         {
-            // ToDo: implement.
-            return array;
+            if (index < 0 || index > array?.Length)
+            {
+                return array;
+            }
+            if (ArrayIsNullOrEmpty(array))
+            {
+                return new int[] { number };
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Codacy Issue found: [Remove the array type; it is redundant.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4930349)

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 6]
```diff
@@ -9,7 +9,58 @@ public static class ArrayOperations
...
+            }
+
+            BubbleSort(array);
+        }
+
+        /// <summary>
+        /// check if an array is null or empty, returning true
+        /// </summary>
+        /// <param name="array">Input array.</param>
+        private static bool ArrayIsNullOrEmpty(int[] array)
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Neat 🙂

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 7]
```diff
@@ -9,7 +9,58 @@ public static class ArrayOperations
         /// <param name="array">Input array in a random order.</param>
         public static void Sort(int[] array)
         {
-            // ToDo: implement.
+            if (ArrayIsNullOrEmpty(array))
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: I love the separation of validation and actual function. 2 high level calls- well done.

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 8]
```diff
@@ -19,18 +70,61 @@ public static void Sort(int[] array)
...
+
+            for (int i = arrayCopy.Length - 1; i >= 0; i--)
+            {
+                array[j] = arrayCopy[i];
+                j++;
+            }
         }
 
+        /// <summary>
+        /// Copy a section of an array in the begining of another array
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: 👍 For summary comments

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 9]
```diff
@@ -39,8 +133,14 @@ public static int[] RemoveLast(int[] array)
...
         public static int[] RemoveFirst(int[] array)
         {
-            // ToDo: implement.
-            return array;
+            if (ArrayIsNullOrEmpty(array))
+            {
+                return array;
+            }
+
+            int[] auxArray = new int[array.Length - 1];
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Reuse the remove at function

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 10]
```diff
@@ -19,18 +70,61 @@ public static void Sort(int[] array)
...
         /// <returns>A new array with the last element removed. If an array is empty or null, returns input array.</returns>
         public static int[] RemoveLast(int[] array)
         {
-            // ToDo: implement.
-            return array;
+            if (ArrayIsNullOrEmpty(array))
+            {
+                return array;
+            }
+
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Reuse remove at function

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           

Complexity increasing per file
==============================
- Src/BootCamp.Chapter1/ArrayOperations.cs  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4930349&bid=16308082)

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 11]
```diff
@@ -19,18 +67,58 @@ public static void Sort(int[] array)
...
         public static int[] RemoveLast(int[] array)
         {
-            // ToDo: implement.
-            return array;
+            if (ArrayIsNullOrEmpty(array))
+            {
+                return array;
+            }
+          
+            return RemoveAt(array, array.Length - 1); ;
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Codacy Issue found: [Remove this empty statement.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4930349)

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- Src/BootCamp.Chapter1/ArrayOperations.cs  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4930349&bid=16308082)

**Comment by ygnacio** [comment]
VS got me a git push error:
`Error encountered while pushing to the remote repository: rejected Updates were rejected because the tip of your current branch is behind its remote counterpart. Integrate the remote changes  before pushing again.`

but pull said nothing to pull, so I did a push force.

**Pull request closed without merging**